src/char/necro.py

```diff
@@ -0,0 +1,524 @@
+import keyboard
+from utils.custom_mouse import mouse
+from char import IChar
+from template_finder import TemplateFinder
+from ui import UiManager
+from pather import Pather
+from logger import Logger
+from screen import Screen
+from utils.misc import wait, rotate_vec, unit_vector
+import random
+from typing import Tuple, Union, List
+from pather import Location, Pather
+import numpy as np
+import time
+from utils.misc import cut_roi, is_in_roi
+
+import mouse as _mouse
+from mouse import _winmouse
+
+class Necro(IChar):
+    def __init__(self, skill_hotkeys, char_config, screen: Screen, template_finder: TemplateFinder, ui_manager: UiManager, pather: Pather):
+        Logger.info('\033[94m'+"Setting up Necro"+'\033[0m')
+        super().__init__(skill_hotkeys, char_config, screen, template_finder, ui_manager)
+        self._pather = pather
+        #custom necro pathing for pindle
+        self._pather.adapt_path((Location.A5_PINDLE_START, Location.A5_PINDLE_SAFE_DIST), [100,101])
+        self._pather.adapt_path((Location.A5_PINDLE_SAFE_DIST, Location.A5_PINDLE_END), [104])
+        #minor offsets to pindle fight locations
+        self._pather.offset_node(102, [15, 0])
+        self._pather.offset_node(103, [15, 0])
+        self._pather.offset_node(101, [100,-5])
+        #custom locations for shenk paths
+        self._pather.adapt_path((Location.A5_SHENK_START, Location.A5_SHENK_SAFE_DIST),[141, 142, 143, 144, 145])
+        self._pather.adapt_path((Location.A5_SHENK_SAFE_DIST, Location.A5_SHENK_END), [149])
+        #custom locations for trav paths
+        self._pather.adapt_path((Location.A3_TRAV_START, Location.A3_TRAV_CENTER_STAIRS), [220, 221, 222, 223, 224, 225, 226])
+
+
+    def _revive(self, cast_pos_abs: Tuple[float, float], spray: int = 10, cast_count: int=12):
+        Logger.info('\033[94m'+"raise revive"+'\033[0m')
+        keyboard.send(self._char_config["stand_still"], do_release=False)
+        for _ in range(cast_count):
+            if self._skill_hotkeys["raise_revive"]:
+                keyboard.send(self._skill_hotkeys["raise_revive"])
+                #Logger.info("revive -> cast")
+            x = cast_pos_abs[0] + (random.random() * 2*spray - spray)
+            y = cast_pos_abs[1] + (random.random() * 2*spray - spray)
+            cast_pos_monitor = self._screen.convert_abs_to_monitor((x, y))
+
+            nx = cast_pos_monitor[0]
+            ny = cast_pos_monitor[1]
+            if(nx>1280):
+                nx=1275
+            if(ny>720):
+                ny=715
```

Comment:
if you are concerned about moving out of the window you should also check for < 0, no?

Reply:
yep should check that as well, ill take a look and see if this is even still needed.